<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Benchmarking Results • ultragsea</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Benchmarking Results">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">ultragsea</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.12</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/ultragsea.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/benchmark.html">Benchmarking Results</a></li>
    <li><a class="dropdown-item" href="../articles/compare-methods.html">Comparing with other methods</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/bigomics/ultragsea/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Benchmarking Results</h1>
                        <h4 data-toc-skip class="author">BigOmics
Analytics</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/bigomics/ultragsea/blob/master/vignettes/benchmark.Rmd" class="external-link"><code>vignettes/benchmark.Rmd</code></a></small>
      <div class="d-none name"><code>benchmark.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="gseabenchmarker">GSEABenchmarkeR<a class="anchor" aria-label="anchor" href="#gseabenchmarker"></a>
</h2>
<p>The GSEABenchmarkeR package (<a href="https://bioconductor.org/packages/release/bioc/html/GSEABenchmarkeR.html" class="external-link">BioC</a>,
<a href="https://github.com/waldronlab/GSEABenchmarkeR" class="external-link">github</a>)
implements an extendable framework for reproducible evaluation of
enrichment methods of gene expression data. It includes a comprehensive
real data compendia (microarray and RNA-seq) of test datasets. Methods
can then be assessed with respect to runtime, statistical significance,
and relevance of the results for the phenotypes investigated.</p>
<p>GSEABenchmarkeR measures the ‘relevance’ of an algorithm by comparing
the significantly detected gene sets against a list of ‘validated
genesets’ in a collection of reference datasets. Somehow this is a
measure of the true positive rate (TPR). Benchmarking was conducted on
42 GEO data sets and a total of 12498 gene sets (GO and KEGG
collections). We ran the following methods through the benchmark:</p>
<ol style="list-style-type: decimal">
<li>UltraC: ultragsea with correlation test.</li>
<li>UltraZ: ultragsea with z-test.</li>
<li>fastFET: a fast implementation of Fisher Exact test (from the
corpora R package) based on the hypergeometric function
<code>phyper</code>.</li>
<li>GOAT: see corresponding R package.</li>
<li>cameraPR: preranked camera from the limma R package.</li>
<li>fGSEA: fast GSEA from the <code>fgsea</code> R package</li>
</ol>
<p>From <strong>Figure 1</strong> we see that ultragsea methods (ultraC
and ultraZ) are the fastest. While in <strong>Figure 2</strong> we can
see that ultraC achieves the highest relevance compared to the
others.</p>
<div class="float">
<img src="fig/benchmark-runtime.png" alt="Figure 1. Run time of enrichment methods (-log10 seconds)."><div class="figcaption">
<strong>Figure 1</strong>. Run time of
enrichment methods (-log10 seconds).</div>
</div>
<div class="float">
<img src="fig/benchmark-relevance-vs-runtime.png" alt="Figure 2. Relevance vs. relative speed-up of enrichment methods. Benchmarking was conducted using the GSEABenchmarkeR package."><div class="figcaption">
<strong>Figure 2</strong>. Relevance
vs. relative speed-up of enrichment methods. Benchmarking was conducted
using the GSEABenchmarkeR package.</div>
</div>
</div>
<div class="section level2">
<h2 id="eabenchmark">EAbenchmark<a class="anchor" aria-label="anchor" href="#eabenchmark"></a>
</h2>
<p>The <a href="https://bitbucket.org/sonnhammergroup/eabenchmark/src/master/" class="external-link">EAbenchmark</a>
R package, presents a generalized benchmark for evaluating enrichment
analysis (EA) methods, utilizing a diverse set of 82 gene expression
datasets across 26 diseases, with a balanced representation of
non-cancer conditions. EAbenchmark introduces the Disease Pathway
Network, which links related KEGG pathways for evaluation, enabling
assessment of the TPR/FPR of EA methods.</p>
<p>EAbenchmark evaluates the false positive rate (FPR) performance of
the EA methods on randomized data (see <strong>Figure 3</strong>). An
ideal method would not report any significantly enriched gene sets.</p>
<p><strong>Figure 4</strong> plot the distribution of p-values of the
methods under the null hypothesis. Enrichment analysis methods should
produce P-values that are uniformly distributed but in practice they are
either biased toward 0 or 1 or exhibit a bimodal distribution biased
toward both extremes. Such a bias can affect the significance of the
analysis, hence we examined the distribution of P-values for each method
to see if it was right- or left-skewed. A right-skewed distribution
(P-values biased toward 0) can potentially lead to false positives by
reporting pathways as impacted when they are not. Conversely, a
left-skewed distribution (P-values biased toward 1) can lead to false
negatives by reporting pathways as not significant when they are
actually impacted.</p>
<p>By plotting true positive rate (TPR) against FPR we can visually
determine the method with highest TPR and lowest FPR (<strong>Figure
5</strong>). We see that GOAT performs the best with similar low FPR but
slightly higher TPR than fGSEA, and the ultraGSEA methods. fastFET and
the single-sample enrichment methods (PLAID and replaidGSVA) had similar
TPR but higher FPR (although still very small, well below 2%).</p>
<p>Finally, <strong>Figure 6</strong> shows the runtimes of the methods.
The ultragsea methods are almost 100x faster than fGSEA and about 10x
faster than cameraPR and the naively implemented ORA Fisher’s exact test
(Fisher). PLAID, our single-sample enrichment method, is still very
fast. The ultragsea methods (ultraGSEA.cor and ultraGSEA.ztest) and
fastFET (hypergeometric-base Fisher’s Exact test) are the fastest.</p>
<div class="float">
<img src="fig/PvalueUnderTrueH0.png" alt="Figure 3. False discovery rate of different enrichment methods under true H0."><div class="figcaption">
<strong>Figure 3</strong>. False discovery rate
of different enrichment methods under true H0.</div>
</div>
<div class="float">
<img src="fig/SkewenessHistogram.png" alt="Figure 4. P-value distribution histograms. Under the null hypothesis, enrichment analysis methods should produce P-values that are uniformly distributed. The colors correspond to the different method classes preranked (blue), ORA based (green) and single-sample based (red)."><div class="figcaption">
<strong>Figure 4</strong>. P-value distribution
histograms. Under the null hypothesis, enrichment analysis methods
should produce P-values that are uniformly distributed. The colors
correspond to the different method classes preranked (blue), ORA based
(green) and single-sample based (red).</div>
</div>
<div class="float">
<img src="fig/TPRvsFPR.png" alt="Figure 5. True positive rate (TPR) vs. false positive rate (1-FPR) of the various enrichment methods. The best methods have high TPR and low FPR."><div class="figcaption">
<strong>Figure 5</strong>. True positive rate
(TPR) vs. false positive rate (1-FPR) of the various enrichment methods.
The best methods have high TPR and low FPR.</div>
</div>
<div class="float">
<img src="fig/RuntimeDataset.png" alt="Figure 6. Relative runtime of the enrichment methods. Time axis is in logscale. The colors correspond to the different method classes preranked (blue), ORA based (green) and single-sample based (red)."><div class="figcaption">
<strong>Figure 6</strong>. Relative runtime of
the enrichment methods. Time axis is in logscale. The colors correspond
to the different method classes preranked (blue), ORA based (green) and
single-sample based (red).</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ivo Kwee, Antonino Zito.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
